### PR TITLE
 bandscope + broadcast - frills = 99.89%

### PR DESCRIPTION
**Pull request by robang74**
Please check this configuration, it fits at the limit:

   text    data     bss     dec     hex filename
  59392      52    3376   62820    f564 f4hwn.bandscope
  57320      20    3012   60352    ebc0 f4hwn.broadcast
  60648      52    3336   64036    fa24 f4hwn.voxless
  61320      52    3336   64708    fcc4 f4hwn.bnb-frills

### CODE REVIEW

**Comment by robang74** [comment]
UPDATE & READY FOR YOUR ATTENTION

   text    data     bss     dec     hex filename
  61308      52    3332   64692    fcb4 f4hwn.fullflash
  59508      52    3372   62932    f5d4 f4hwn.bandscope
  57424      20    3008   60452    ec24 f4hwn.broadcast
  60764      52    3332   64148    fa94 f4hwn.voxless

**Comment by armel** [comment]
Hi. It's probably interesting for your needs (no AirCopy, no FlashLight, no Copy Chan to VFO, with PMR what is useless since TXLock add, etc.). Now, sorry, but I don't intend to add everyone's custom configurations. Otherwise, we'll never see the end of it. 

Everyone has their own vision of their needs. My job is to provide a few generic versions. But I wouldn't go beyond that. It's up to each individual to adapt the code and recompile if necessary. That's why my firmware is open source.

**Comment by robang74** [comment]
Well, my personal configuration can be ignored but shrinking the code is another story that it is useful for everybody. Flashlight entered into the FW. No AirCopy is the same with your configuration Broadcat + Bandscope v.<3.0. About these two: no Copy Chan to VFO, with PMR what is useless since TXLock add - I have no a practical answer about. 

Also the Flashlight when F4HWN is involved is quite limited compared with the factory. After all, my mod is designed to have a cheap home radio station for listening and get spectrum. Under this point of view, also flashlight is quite useless (no outdoor activities). By the way, I have generalized a little the script and now the configuration are WAY to easy for doing a confrontation.

I think you have to be more positive, also with other contributors. It is meaningless HOW you would like to name some constant. It is a task than can better described with a reg-expression and then feed sed or awk. People are here for have fun not to please your will to rename things... ;-)

Now, that I gave you also a suggestion about how to deal with contributors, can you explain yourself in more extensive way about "no Copy Chan to VFO, with PMR what is useless since TXLock add" where Chan is channel and VFO is the frequency mode AFAIK. About PMR, I was thinking that it was a Power average of something but I saw googling that it is probably a group of channels (and then, I am not interested in). I have no clue about TX lock what's referring about. Frequency lock or transmission lock. Because in the 2nd case, it can be unlock unless the hardware prevents it.

= Robang74's configuraiton =

make_in_docker "\
ENABLE_FLASHLIGHT=1 \
ENABLE_SPECTRUM=1 \
ENABLE_FMRADIO=1 \
ENABLE_VOX=0 \
ENABLE_AIRCOPY=0 \
ENABLE_AUDIO_BAR=0 \
ENABLE_FEAT_F4HWN_SLEEP=0 \
ENABLE_COPY_CHAN_TO_VFO=0 \
ENABLE_FEAT_F4HWN_SPECTRUM=1 \
ENABLE_FEAT_F4HWN_RX_TX_TIMER=0 \
ENABLE_FEAT_F4HWN_PMR=1 \
TARGET=f4hwn.fullflash \
&& cp -f f4hwn.fullflash* compiled-firmware/"

= Original Configurations =

make_in_docker "\
ENABLE_SPECTRUM=1 \
ENABLE_FMRADIO=0 \
ENABLE_AIRCOPY=1 \
TARGET=f4hwn.bandscope \
&& cp -f f4hwn.bandscope* compiled-firmware/"

make_in_docker "\
ENABLE_SPECTRUM=0 \
ENABLE_FMRADIO=1 \
ENABLE_AIRCOPY=1 \
TARGET=f4hwn.broadcast \
&& cp -f f4hwn.broadcast* compiled-firmware/"

make_in_docker "\
ENABLE_SPECTRUM=1 \
ENABLE_FMRADIO=1 \
ENABLE_VOX=0 \
ENABLE_AIRCOPY=0 \
ENABLE_AUDIO_BAR=0 \
ENABLE_FEAT_F4HWN_SPECTRUM=0 \
ENABLE_FEAT_F4HWN_SLEEP=0 \
TARGET=f4hwn.voxless \
&& cp -f f4hwn.voxless* compiled-firmware/"

**Comment by robang74** [comment]
Sorry, I am not interested to get personal and have a debate about. You got the message and touched your soul, I hope for the better. This is enough for me. Be back to the practicing. You are right, I have started today precisely. And unless I mess-up things (which is possible as much as I can fix them because the states logic is not hard to handle), I have my reason to see quite a thing here.

UPDATE ABOUT SIZE

64692 - 64668 = PMR weights 24 bytes on the total size. Few, a lot? Everything is relative. In the last year, the two commits that proudly showed a reduction in size were about 40 and 80 bytes. Hence 24 bytes is a relatively important amount. Obviously, functionalities that are not useful in a certain configuration are superfluous.

This is the last version of your main:

   text    data     bss     dec     hex filename
  60856      52    3388   64296    fb28 f4hwn.bandscope
  58836      20    3032   61888    f1c0 f4hwn.broadcast
  61016      52    3336   64404    fb94 f4hwn.voxless
 
This is the last version of my main:

   text    data     bss     dec     hex filename
  59500      52    3372   62924    f5cc f4hwn.bandscope
  57416      20    3008   60444    ec1c f4hwn.broadcast
  60756      52    3332   64140    fa8c f4hwn.voxless
 
The differences are quite important, IMHO:

   1356       0      16     1372 f4hwn.bandscope
   1420       0      24     1444 f4hwn.broadcast
    260       0       4      264 f4hwn.voxless
   
I have reduced the size of 264 bytes in the worst case and 1372 in the best case. Considering that I started to work on this firmware today (8h, ago) for the first time in my life in dealing with a ham radio, it seems to me quite a thing. Right? However, I suggest you check my changes. I did them for myself but you can imagine, it is still a work of someone that started today, literally.

File: compile-with-docker-all.sh [hunk 1]
```diff
@@ -1,30 +1,61 @@
-#!/bin/sh
+#!/bin/bash
 #export DOCKER_DEFAULT_PLATFORM=linux/amd64
+
 IMAGE_NAME="uvk5"
-rm "${PWD}/compiled-firmware/*"
+DEST_DIR="compiled-firmware/"
+LOCL_DIR="${PWD}/${DEST_DIR}/"
+ALWAYSIN="ENABLE_NOAA=0 NABLE_UART=1"
```
Comment (prokrypt) [inline compile-with-docker-all.sh:7]: NABLE_UART typo

Comment (robang74) [reply compile-with-docker-all.sh:7]: Thanks, a mistake that can cost a bricked device! Isn't it, right?

Comment (prokrypt) [reply compile-with-docker-all.sh:7]: Thankfully these devices are pretty hard to brick. UART is not 100% needed for normal operation unless you want to use Chirp, etc.

**Comment by robang74** [comment]
Just 8 bytes more with UART enabled, does it sound reasonable?

  text    data     bss     dec     hex filename
  61308      52    3332   64692    fcb4 f4hwn.fullflash
  59508      52    3372   62932    f5d4 f4hwn.bandscope
  57424      20    3008   60452    ec24 f4hwn.broadcast
  60764      52    3332   64148    fa94 f4hwn.voxless

**Comment by prokrypt** [comment]
It was probably already enabled? ¯\\_(ツ)\_/¯

**Comment by robang74** [comment]
> It was probably already enabled? ¯\_(ツ)_/¯

I hope so, because if UART means the thing that let people use the cable to upload another firmware...

...it is crazy the idea to disable it by uploading a firmware! Not all that can be disabled, it should be disabled.

Or probably, enabling the UART is just a single write on the eeprom at the boot. I have to check, I didn't yet.

**Comment by prokrypt** [comment]
> > It was probably already enabled? ¯_(ツ)_/¯
> 
> I hope so, because if UART means the thing that let people use the cable to upload another firmware...
> 
> ...it is crazy the idea to disable it by uploading a firmware! Not all that can be disabled, it should be disabled.
> 
> Or probably, enabling the UART is just a single write on the eeprom at the boot. I have to check, I didn't yet.

Firmware flashing is handled by the radio's bootloader which is never overwritten. UART is only for serial communication (radio programming, control, etc) after the firmware has booted.

**Comment by armel** [comment]
> This is the last version of your main:
> 
> text data bss dec hex filename
> 60856 52 3388 64296 fb28 f4hwn.bandscope
> 58836 20 3032 61888 f1c0 f4hwn.broadcast
> 61016 52 3336 64404 fb94 f4hwn.voxless
> 
> This is the last version of my main:
> 
> text data bss dec hex filename
> 59500 52 3372 62924 f5cc f4hwn.bandscope
> 57416 20 3008 60444 ec1c f4hwn.broadcast
> 60756 52 3332 64140 fa8c f4hwn.voxless

Because you're not comparing the same things.
For example, your bandscope version disables the following functions: 

ENABLE_VOX = 0
ENABLE_AUDIO_BAR = 0
ENABLE_COPY_CHAN_TO_VFO = 0
ENABLE_FEAT_F4HWN_RX_TX_TIMER = 0

The real difference is probably about 40 bytes (due to your flags.h optimization, which is probably a good thing). That's all... 40 bytes, not 1 372.

**Comment by robang74** [comment]
> The real difference is probably about 40 bytes (due to your flags.h optimization, which is probably a good thing). That's all... 40 bytes, not 1 372.

Correct. I changed also the Makefile which creates a certain impact also in the original configurations. Hard to say, probably 48.

Except for one configuration, in which bandscope (the new 3.x not the old) and the broadcast can meet together and something more like flashlight, etc. Obviously not VOX and other complicated feature like VOX which recognize the voice and open the channel to TX without pressing the PTT button. Except for that configuration, for those are asking more and more...

Screenshot from 2024-10-21 21-36-01

**Comment by robang74** [comment]
And probably another -30 bytes in this commit that it still in "devel" branch. :-)

https://github.com/robang74/uv-k5-firmware-custom/commit/8bb54ef9847136ee1339d7079bfaa432dafc8c95

**Comment by armel** [comment]
Hello,

Well... I see that your PR changes the labels and the order of the menus. Regarding the order, it's important to note that it follows a specific logic, especially related to channel attributes. Were you aware of this? I don't think so. 

I should add that the numbers are sometimes familiar to users. I want to go to ChSave, I enter the Menu by pressing M and type 17 on the keyboard...

As for the labels, it's essential to know that users have been familiar with them for over a year (Dual Tachyon > OneOfEleven > Egzumer > F4HWN). Changing them for personal convenience would require updating documentation in multiple languages (English, French, etc.), which doesn't seem justified in this case.

Additionally, with the same build parameters (Makefile), your binary is larger than mine. 

In short, I don't understand what your objective is.

I would also like to emphasize the importance of [atomic PR](https://medium.com/@chikim79/the-art-of-creating-atomic-prs-16d13b3828e3), which is a good practice that would help manage changes more effectively.

Therefore, I will close this PR as it diverges from the direction the project has been taking for the past year. You'll probably take offence, given your preliminary judgment of me. I wish you the best of luck in your further development. And have fun.

**Comment by robang74** [comment]
> Hello,

Sorry, I forgot this pull open and I continued to push on my main. About everything else, who cares? Cherry-picking, seek for inspiration. Open Source is about learning from others people code not debating about bureaucratic practices. You like it, take it. Do not like it? Rewrite. Moreover, I am not paid to listen your *bossy" comments nor even to following it or working for you. Software libre like free speech not like free beer. It is time to return the old fashion style of hacking. ;-)

Best regards, R-
